src/hotspot/os/linux/os_linux.cpp

```diff
@@ -6019,8 +6021,143 @@ class CracSHM {
   }
 };
 
+class Freeze {
+  static const int signo = SIGUSR1;
+  struct sigaction act_old;
+
+  template<class Callback> bool all_threads(Callback callback) const {
+    const char dirname[] = "/proc/self/task";
+    GrowableArray<pid_t> done;
+    bool signalled;
+    do {
+      signalled = false;
+      DIR *dir = opendir(dirname);
+      if (dir == NULL) {
+        tty->print_cr("JVM: Error opening %s: %m", dirname);
+        return false;
+      }
+      pid_t tid = gettid();
+      struct dirent *ent;
+      while (errno = 0, ent = readdir(dir)) {
+        const char *name = ent->d_name;
+        if (strcmp(name, ".") == 0)
+          continue;
+        if (strcmp(name, "..") == 0)
+          continue;
+        char *endptr;
+        unsigned long l = strtoul(name, &endptr, 10);
+        pid_t ent_tid = l;
+        if (l >= LONG_MAX || ent_tid != (long) l) {
+          tty->print_cr("JVM: Error parsing %s entry \"%s\"", dirname, name);
+          return false;
+        }
+        if (ent_tid == tid)
+          continue;
+        if (!done.append_if_missing(ent_tid))
+          continue;
+        signalled = true;
+        callback(ent_tid);
+      }
+      if (errno) {
+        tty->print_cr("JVM: Error reading %s: %m", dirname);
+        return false;
+      }
+      if (closedir(dir)) {
+        tty->print_cr("JVM: Error closing %s: %m", dirname);
+        return false;
+      }
+    } while (signalled);
+    return true;
+  }
+  static pthread_mutex_t signalled_mutex;
+  static pthread_cond_t signalled_cond;
+  static size_t signalled;
+  static pthread_cond_t resume_cond;
+  static void handler(int handler_signo) {
+    assert(handler_signo == signo, "handler signo");
+    int err;
+    err = pthread_mutex_lock(&signalled_mutex);
+    assert(!err, "pthread error");
+    ++signalled;
+    err = pthread_mutex_unlock(&signalled_mutex);
+    assert(!err, "pthread error");
+    err = pthread_cond_signal(&signalled_cond);
+    assert(!err, "pthread error");
+    pthread_mutex_t unused_mutex = PTHREAD_MUTEX_INITIALIZER;
```

Comment:
Waiting on a local mutex deserves a comment to explain the approach.  I think we're waiting until the thaw operation signals the resume condition but better to be explicit in the code on the intended interactions between the waiters

Reply:
Added: `// Just wait until thaw() is called. No mutex is needed for that.`